core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -244,6 +244,31 @@ class SparkHadoopUtil extends Logging {
   }
 
   /**
+   * List directories/files matching the path and return the `FileStatus` results.
+   * If the pattern is not a regexp then a simple `getFileStatus(pattern)`
+   * is called to get the status of that path.
+   * If the path/pattern does not match anything in the filesystem,
+   * an empty sequence is returned.
+   * @param pattern pattern
+   * @return a possibly empty array of FileStatus entries
+   */
+  def globToFileStatus(pattern: Path): Array[FileStatus] = {
```

Comment:
I recognise your concerns.

# the actual check is the one used already in {{globPathIfNecessary}}; that's used in {{DataSource}} and {{FileStreamSource}} and is pretty much the main way wildcards are used to locate files. If it was broken, things would have surfaced.
# its essentially an optimisation to bypass a bit of code which, if you looked at its internals, isn't very efficient. The check for expansion characters is pessimistic, to avoid picking up on wildcard patterns.
# ...but, there are no standalone tests for {{globPathIfNecessary}} —so it may be that there are some failure modes that haven't yet surfaced; regressions waiting to happen.

Given it's a less significant optimisation than dodging all the getFileStatus calls, how about I split that into its own patch, something which should also add those checks for {{globPathIfNecessary}}? I'll trim this patch down to the changes in the streaming package —production code and associated tests